app/src/main/java/tech/relaycorp/gateway/common/Operation.kt

```diff
@@ -0,0 +1,9 @@
+package tech.relaycorp.gateway.common
+
+sealed class Operation<out T> {
+    data class Success<T>(val result: T) : Operation<T>()
+    data class Error<T>(val throwable: Throwable) : Operation<T>()
+}
+
+val <T> Operation<T>.isSuccessful get() = this is Operation.Success<T>
+val <T> Operation<T>.isError get() = this is Operation.Error<T>
```

Comment:
Out of curiosity as someone new to Kotlin/Android, why use this class vs simply raising exceptions when things go wrong?

Reply:
It's a good question, and I thought about it myself too. This Operation pattern makes it more clear that the error case should be handled. The sealed class just helps for the kotlin `when`. This is the reactive programming pattern, not far away for the `Either` pattern: https://www.ibm.com/developerworks/library/j-ft13/index.html

But in practice, there's not a huge difference.

Reply:
:+1: Thanks for the explanation!